docs/long_term_roadmap.md

```diff
@@ -0,0 +1,250 @@
+- The irregular support surface area of the large number of PyTorch ops across
+  the Linalg, TOSA, and MHLO backends.
+
+Most of this document describes long-term ecosystem changes that will address
+these, drastically improving Torch-MLIR's ability to meet its goals.
+
+## Roadmap
+
+### Refactoring the frontend
+
+The primary way to make the frontend more reliable is to leverage new PyTorch
+infrastructure that bridges from the PyTorch eager world into compiler-land.
+PyTorch has two main projects that together cover almost all user use cases and
+provide a technically sound, high quality-of-implementation path from user
+programs into the compiler.
+
+- [TorchDynamo](https://github.com/pytorch/torchdynamo) - TorchDynamo uses
+  tracing-JIT-like techniques and program slicing to extract traces of tensor
+  operations, which can then be passed to lower-level compilers. It works
+  seamlessly with unmodified user programs.
+- [FuncTorch](https://github.com/pytorch/functorch) - FuncTorch is basically JAX
+  for PyTorch. It requires manual program tracing and slicing, but that is
+  actually important for users since it gives them direct control over various
+  important transformations, such as `grad` and `vmap`.
+
+These are both being heavily-invested-in by PyTorch core developers, and are
+generally seen as the next generation of compiler technology for the project,
+blending PyTorch's famous usability with excellent compiler integration
+opportunities. Torch-MLIR works with these technologies as they exist today, but
+significant work remains to enable wholesale deleting the high-maintenance parts
+of Torch-MLIR. In the future, we expect the block diagram of Torch-MLIR to be
+greatly simplified, as shown in the diagram below. Note that in the "Future"
+side, PyTorch directly gives us IR in a form satisfying the backend contract.
+
+![Roadmap of the frontend](images/roadmap_frontend.png)
+
+The primary functional requirement of Torch-MLIR which remains unaddressed by
+today's incarnation of TorchDynamo and FuncTorch is the support for dynamic
+shapes. PyTorch core devs are
+[heavily investing](https://dev-discuss.pytorch.org/t/state-of-symbolic-shapes-branch/777)
+in this area, and both TorchDynamo and FuncTorch are being upgraded as PyTorch
+rolls out its new symbolic shape infrastructure.
+
+Smaller blockers are related to general API stability and usability of the
+various pieces of PyTorch infra.
+
+These blockers are expected to be addressed by the PyTorch core devs over time.
+Torch-MLIR's role here is to communicate our requirements to PyTorch core and
+align their roadmap and ours. We do this by maintaining connections with the
+PyTorch core developers and being "good-citizen power users" of their latest
+technology (i.e. trying things out, surfacing bugs, providing feedback, etc.).
+
+Note: Because both TorchDynamo and FuncTorch are TorchFX-based, we could write a
```

Comment:
TorchFX lacks support for dynamic control flow, but TorchScript does. Deleting the TorchScript importer may prevent backend compilers from seeing control flow, if they could do any optimization.

Reply:
I don't think he means to drop the top-level TorchScript importer; he probably means the lowering path torch.fx -> TorchScript -> MLIR. We should deprecate it in favor for direct lowering from torch.fx -> MLIR.